File: packages/python/tests/test_rest.py
from __future__ import annotations

import json

import httpx
import pytest

from erpc_sdk import (
    ErpcClient,
    ErpcClientConfig,
    ErpcCloudClient,
    ErpcCloudClientConfig,
    ErpcConfigError,
    ErpcInvalidResponseError,
)


@pytest.mark.asyncio
async def test_price_rest_uses_bearer_and_repeated_ids() -> None:
    def handler(request: httpx.Request) -> httpx.Response:
        assert request.headers["authorization"] == "Bearer secret"
        assert request.url.path == "/v2/updates/price/latest"
        assert request.url.params.get_list("ids[]") == ["feed-a", "feed-b"]
        assert request.url.params["parsed"] == "true"
        return httpx.Response(200, json={"binary": {"data": ["value"], "encoding": "base64"}})

    http = httpx.AsyncClient(transport=httpx.MockTransport(handler))
    erpc = ErpcClient(
        ErpcClientConfig("secret", endpoint="https://example.test"),
        http_client=http,
    )
    result = await erpc.price.get_latest_price_updates(["feed-a", "feed-b"], parsed=True)
    assert result["binary"]["encoding"] == "base64"
    await erpc.close()
    await http.aclose()


@pytest.mark.asyncio
async def test_price_sse_parses_event_fields_and_multiline_data() -> None:
    update = {"binary": {"data": [], "encoding": "hex"}}
    body = f"id: 7\nevent: price_update\ndata: {json.dumps(update)}\n\n"

    def handler(_: httpx.Request) -> httpx.Response:
        return httpx.Response(200, content=body.encode())

    http = httpx.AsyncClient(transport=httpx.MockTransport(handler))
    erpc = ErpcClient(ErpcClientConfig("secret"), http_client=http)
    events = [event async for event in erpc.price.stream_price_updates(["feed"])]
    assert events == [{"id": "7", "event": "price_update", "data": update}]
    await erpc.close()
    await http.aclose()


@pytest.mark.asyncio
async def test_usage_validates_year_month_before_transport() -> None:
    http = httpx.AsyncClient(transport=httpx.MockTransport(lambda _: httpx.Response(500)))
    erpc = ErpcClient(ErpcClientConfig("secret"), http_client=http)
    with pytest.raises(ErpcConfigError, match="YYYY-MM"):
        await erpc.usage.get_monthly_api_key_usage("2026-13")
    await erpc.close()
    await http.aclose()


@pytest.mark.asyncio
async def test_usage_unwraps_and_validates_envelope() -> None:
    usage = {
        "apiKeys": [],
        "chains": [],
        "hasStrandedUsage": False,
        "keyCount": 0,
        "totalCount": 0,
        "totalCredits": 0,
        "updatedAt": None,
        "yearMonth": "2026-08",
    }

    def handler(request: httpx.Request) -> httpx.Response:
        assert request.url.params["yearMonth"] == "2026-08"
        return httpx.Response(200, json={"success": True, "message": usage})

    http = httpx.AsyncClient(transport=httpx.MockTransport(handler))
    erpc = ErpcClient(ErpcClientConfig("secret"), http_client=http)
    assert await erpc.usage.get_monthly_api_key_usage("2026-08") == usage
    await erpc.close()
    await http.aclose()


def test_cloud_rejects_non_local_plain_http_and_redacts_token() -> None:
    with pytest.raises(ErpcConfigError, match="HTTPS"):
        ErpcCloudClientConfig("token", endpoint="http://example.test")
    config = ErpcCloudClientConfig(" top-secret ", endpoint="http://localhost:9000/")
    assert "top-secret" not in repr(config)
    assert config.endpoint == "http://localhost:9000/"


@pytest.mark.asyncio
async def test_cloud_catalog_and_resource_paths_are_scoped_reads() -> None:
    paths: list[str] = []

    def handler(request: httpx.Request) -> httpx.Response:
        paths.append(request.url.path)
        assert request.headers["authorization"] == "Bearer access"
        if request.url.path.endswith("/catalog"):
            return httpx.Response(
                200,
                json={
                    "success": True,
                    "message": {
                        "offerings": [
                            {
                                "id": "vps-small",
                                "kind": "vps",
                                "name": "Small",
                                "description": "Small compute",
                                "regions": ["eu"],
                                "capabilities": ["compute"],
                            }
                        ]
                    },
                },
            )
        return httpx.Response(
            200,
            json={
                "success": True,
                "message": {"resource": {"id": "a/b", "kind": "vps", "status": "ready"}},
            },
        )

    http = httpx.AsyncClient(transport=httpx.MockTransport(handler))
    cloud = ErpcCloudClient(
        ErpcCloudClientConfig("access", endpoint="https://example.test/base"),
        http_client=http,
    )
    assert (await cloud.catalog.list())[0]["id"] == "vps-small"
    assert (await cloud.resources.get("a/b"))["status"] == "ready"
    assert paths == ["/base/v4/cloud/catalog", "/base/v4/cloud/resources/a/b"]
    await cloud.close()
    await http.aclose()


@pytest.mark.asyncio
async def test_invalid_account_response_is_rejected() -> None:
    http = httpx.AsyncClient(
        transport=httpx.MockTransport(lambda _: httpx.Response(200, json={"plan": "unknown"}))
    )
    erpc = ErpcClient(ErpcClientConfig("secret"), http_client=http)
    with pytest.raises(ErpcInvalidResponseError):
        await erpc.account.get_token_balance()
    await erpc.close()
    await http.aclose()

File: packages/ruby/test/test_rest.rb
# frozen_string_literal: true

require_relative "test_helper"

class RestTest < Minitest::Test
  def test_keyless_price_sse_fails_locally_without_http_io
    calls = 0
    adapter = FakeHttpAdapter.new do
      calls += 1
      raise "keyless SSE should not use HTTP"
    end
    erpc = ERPC::Client.new(
      ERPC::ClientConfig.new(
        ethereum_rpc: ERPC::RpcEndpointConfig.new(http_url: "https://customer.example/rpc")
      ),
      http_adapter: adapter
    )

    assert_raises(ERPC::NotConfiguredError) do
      erpc.price.stream_price_updates(ids: ["feed"]).to_a
    end
    assert_equal 0, calls
    assert_empty adapter.requests
  ensure
    erpc&.close
  end

  def test_price_rest_uses_bearer_and_repeated_ids
    adapter = FakeHttpAdapter.new do |request|
      assert_equal "Bearer secret", request[:headers]["authorization"]
      uri = URI.parse(request[:url])
      assert_equal "/v2/updates/price/latest", uri.path
      assert_equal ["feed-a", "feed-b"], URI.decode_www_form(uri.query).filter_map { |key, value| value if key == "ids[]" }
      assert_includes URI.decode_www_form(uri.query), ["parsed", "true"]
      ERPC::HttpResponse.new(
        status: 200,
        body: JSON.generate("binary" => { "data" => ["value"], "encoding" => "base64" })
      )
    end
    erpc = ERPC::Client.new(
      ERPC::ClientConfig.new(api_key: "secret", endpoint: "https://example.test"),
      http_adapter: adapter
    )

    result = erpc.price.get_latest_price_updates(ids: %w[feed-a feed-b], parsed: true)
    assert_equal "base64", result.dig("binary", "encoding")
  ensure
    erpc&.close
  end

  def test_price_sse_parses_event_fields_and_data
    update = { "binary" => { "data" => [], "encoding" => "hex" } }
    adapter = FakeHttpAdapter.new { raise "ordinary request should not be used" }
    adapter.stream_chunks = ["id: 7\nevent: price_update\ndata: #{JSON.generate(update)}\n", "\n"]
    erpc = ERPC::Client.new(ERPC::ClientConfig.new(api_key: "secret"), http_adapter: adapter)

    assert_equal(
      [{ "id" => "7", "event" => "price_update", "data" => update }],
      erpc.price.stream_price_updates(ids: ["feed"]).to_a
    )
  ensure
    erpc&.close
  end

  def test_usage_validates_year_month_before_transport
    adapter = FakeHttpAdapter.new { raise "network should not be reached" }
    erpc = ERPC::Client.new(ERPC::ClientConfig.new(api_key: "secret"), http_adapter: adapter)

    assert_raises(ERPC::ConfigError) { erpc.usage.get_monthly_api_key_usage("2026-13") }
    assert_empty adapter.requests
  ensure
    erpc&.close
  end

  def test_usage_unwraps_and_validates_envelope
    usage = {
      "apiKeys" => [],
      "chains" => [],
      "hasStrandedUsage" => false,
      "keyCount" => 0,
      "totalCount" => 0,
      "totalCredits" => 0,
      "updatedAt" => nil,
      "yearMonth" => "2026-08"
    }
    adapter = FakeHttpAdapter.new do |request|
      assert_includes URI.decode_www_form(URI.parse(request[:url]).query), ["yearMonth", "2026-08"]
      ERPC::HttpResponse.new(status: 200, body: JSON.generate("success" => true, "message" => usage))
    end
    erpc = ERPC::Client.new(ERPC::ClientConfig.new(api_key: "secret"), http_adapter: adapter)

    assert_equal usage, erpc.usage.get_monthly_api_key_usage("2026-08")
  ensure
    erpc&.close
  end

  def test_cloud_rejects_non_local_http_and_redacts_token
    assert_raises(ERPC::ConfigError) do
      ERPC::CloudClientConfig.new(access_token: "token", endpoint: "http://example.test")
    end
    config = ERPC::CloudClientConfig.new(access_token: " top-secret ", endpoint: "http://localhost:9000/")
    refute_includes config.inspect, "top-secret"
    assert_equal "http://localhost:9000/", config.endpoint
  end

  def test_cloud_catalog_and_resource_paths_are_scoped_reads
    paths = []
    adapter = FakeHttpAdapter.new do |request|
      uri = URI.parse(request[:url])
      paths << uri.path
      assert_equal "Bearer access", request[:headers]["authorization"]
      if uri.path.end_with?("/catalog")
        body = {
          "success" => true,
          "message" => {
            "offerings" => [{
              "id" => "vps-small",
              "kind" => "vps",
              "name" => "Small",
              "description" => "Small compute",
              "regions" => ["eu"],
              "capabilities" => ["compute"]
            }]
          }
        }
      else
        body = {
          "success" => true,
          "message" => { "resource" => { "id" => "a/b", "kind" => "vps", "status" => "ready" } }
        }
      end
      ERPC::HttpResponse.new(status: 200, body: JSON.generate(body))
    end
    cloud = ERPC::CloudClient.new(
      ERPC::CloudClientConfig.new(access_token: "access", endpoint: "https://example.test/base"),
      http_adapter: adapter
    )

    assert_equal "vps-small", cloud.catalog.list.first["id"]
    assert_equal "ready", cloud.resources.get("a/b")["status"]
    assert_equal ["/base/v4/cloud/catalog", "/base/v4/cloud/resources/a%2Fb"], paths
  end

  def test_invalid_account_response_is_rejected
    adapter = FakeHttpAdapter.new do
      ERPC::HttpResponse.new(status: 200, body: JSON.generate("plan" => "unknown"))
    end
    erpc = ERPC::Client.new(ERPC::ClientConfig.new(api_key: "secret"), http_adapter: adapter)

    assert_raises(ERPC::InvalidResponseError) { erpc.account.get_token_balance }
  ensure
    erpc&.close
  end
end
